source/common/network/lc_trie.h

```diff
@@ -0,0 +1,487 @@
+#pragma once
+
+#include <arpa/inet.h>
+
+#include <algorithm>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/network/address.h"
+
+#include "common/network/cidr_range.h"
+#include "common/network/utility.h"
+
+#include "absl/numeric/int128.h"
+#include "fmt/format.h"
+
+namespace Envoy {
+namespace Network {
+namespace LcTrie {
+
+/**
+ * Level Compressed Trie implemenation for IP(Ipv4 or IPv6) Prefixes.
+ */
+class LcTrie {
+public:
+  /**
+   * @param fill_factor supplies the fraction of completeness to use when considering
+   *        if a subtrie should be compressed.
+   * @param root_branching_factor supplies the branching factor at the root. This is
+   *        an optimization as suggested in the paper.
+   * @param tag_data supplies the ip tag and the prefixes for associated with that tag.
+   */
+  LcTrie(const std::vector<std::pair<std::string, std::vector<Address::CidrRange>>>& tag_data,
+         double fill_factor = 0.5, uint32_t root_branching_factor = 0);
+
+  /**
+   * Search the LC Trie for a prefix that has the input, ip, in range.
+   * @param  ip supplies the IP address searching for a prefix in string format.
+   * @return tag from the prefix that encompasses the input. An empty string is returned
+   *         if no prefix in the LC Trie exists.
+   */
+  std::string search(Network::Address::InstanceConstSharedPtr ip_address) const;
+
+private:
+  /**
+   * Extract n bits from input starting at position p.
+   * @param position supplies the position.
+   * @param n supplies the number of bits to extract.
+   * @param input supplies the
+   */
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)>
+  static IpType extractBits(int p, int n, IpType input) {
+    // The IP's are stored in network byte order. By shifting the value to the left by p bits, the
+    // bits between 0 and p-1 are zero'd out. Then to get the n bits, shift the IP back by the
+    // address_size minus the number of desired bits.
+    return input << p >> (address_size - n);
+  }
+
+  /**
+   * Removes n bits from input starting at 0.
+   * @param n supplies the number of bits to remove.
+   * @param input supplies the
+   */
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)>
+  static IpType removeBits(int n, IpType input) {
+    // The IP's are stored in network byte order. By shifting the value to the left by n bits and
+    // back, this zeros out all bits between 0 and n-1.
+    return input << n >> n;
+  }
+
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)> struct IpPrefixes {
```

Comment:
Should this struct be named "IpPrefix," singular, since it holds just one prefix?

Reply:
Yes, that is clearer. I've updated it to be IpPrefix.